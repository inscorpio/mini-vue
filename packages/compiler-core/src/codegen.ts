import { isString, nullable } from '@mini-vue/shared'
import { NodeTypes } from './ast'
import { CREATE_ELEMENT_VNODE, TO_DISPLAY_STRING, helperMapName } from './runtimeHelper'

export function generate(ast) {
  const { codegenNode } = ast
  const context = createCodegenContext()
  const { push } = context
  const functionName = 'render'
  const args = ['_ctx']
  const signature = args.join(', ')
  genFunctionPreamble(ast, context)
  push(`return function ${functionName}(${signature}) {`)
  push('\n')
  push('  return ')
  genNode(codegenNode, context)
  push('\n')
  push('}')
  return {
    code: context.code,
  }
}

function genFunctionPreamble(ast, context) {
  const { helpers } = ast
  const { push } = context
  const VueBinding = 'Vue'
  const aliasHelper = (s) => {
    const helper = helperMapName[s]
    return `${helper}: _${helper}`
  }

  if (helpers.size) {
    push(`const { ${Array.from(helpers).map(aliasHelper).join(', ')} } = ${VueBinding}\n`)
  }
}

function genNode(node, context) {
  const { push } = context
  if (isString(node)) {
    push(node)
    return
  }
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break
    case NodeTypes.COMPOUND_EXPRESSION:
      genChildren(node.children, context)
      break
    default:
      break
  }
}

function genChildren(children, context) {
  children.forEach(child => genNode(child, context))
}

function genElement(node, context) {
  const { push, helper } = context
  const { tag, props, children } = node
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  push(`"${tag}"`)
  if (children.length) {
    push(', ')
    push(`${nullable(props)}`)
    push(', ')
    if (children.length > 1) {
      push('[')
      genNodeList(children, context)
      push(']')
    }
    else {
      genChildren(children, context)
    }
  }
  else if (props) {
    push(props)
  }
  push(')')
}

function genNodeList(nodes, context) {
  const { push } = context
  nodes.forEach((node, i) => {
    genNode(node, context)
    if (i < nodes.length - 1) {
      push(', ')
    }
  })
}

function genExpression(node, context) {
  const { push } = context
  push(node.content)
}

function genInterpolation(node, context) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genExpression(node.content, context)
  push(')')
}

function genText(node, context) {
  const { push } = context
  push(`"${node.content}"`)
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    },
  }
  return context
}
