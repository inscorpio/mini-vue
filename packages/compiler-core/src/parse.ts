import { NodeTypes } from './ast'

const startDelimiter = '{{'
const endDelimiter = '}}'

export function baseParse(content: string) {
  const context = createParseContext(content)
  return createRoot(context)
}

function createRoot(context: { source: string }) {
  const children = parseChildren(context)
  return {
    children,
  }
}

function createParseContext(content: string) {
  return {
    source: content,
  }
}

function parseChildren(context) {
  const nodes = []

  // 没解析完就一直解析
  while (context.source) {
    let node

    if (context.source.startsWith(startDelimiter)) {
      node = parseInterpolation(context)
    }

    nodes.push(node)
  }
  return nodes
}

function parseInterpolation(context) {
  // {{message}}
  const index = context.source.indexOf(endDelimiter)

  // 解析 {{
  advanceBy(context, startDelimiter.length)
  const content = context.source.slice(0, index - startDelimiter.length)
  // 解析 content
  advanceBy(context, content.length)
  // 解析 }}
  advanceBy(context, endDelimiter.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content.trim(),
    },
  }
}

function advanceBy(context, length) {
  context.source = context.source.slice(length)
}
