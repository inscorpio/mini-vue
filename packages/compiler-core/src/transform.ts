import { NodeTypes } from './ast'
import { TO_DISPLAY_STRING } from './runtimeHelper'

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  traverseNode(root, context)
  createRootCodegen(root)
  createRootHelpers(root, context)
}

function createRootHelpers(root, context) {
  const { helpers } = context
  root.helpers = helpers
}

function createRootCodegen(root) {
  const child = root.children[0]
  root.codegenNode = child
}

function traverseNode(node, context) {
  const { nodeTransforms, helpers } = context
  nodeTransforms.forEach(nodeTransform => nodeTransform(node, context))
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      helpers.add(TO_DISPLAY_STRING)
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
    case NodeTypes.COMPOUND_EXPRESSION:
      traverseChildren(node, context)
      break
    default:
      break
  }
}

function traverseChildren(parent, context) {
  const { children = [] } = parent
  children.forEach(node => traverseNode(node, context))
}

function createTransformContext(
  root,
  {
    nodeTransforms = [],
  },
) {
  return {
    root,
    nodeTransforms,
    helpers: new Set(),
  }
}
