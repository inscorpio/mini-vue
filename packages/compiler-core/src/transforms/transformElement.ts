import { NodeTypes } from '../ast'
import { CREATE_ELEMENT_VNODE } from '../runtimeHelper'

export function transformElement(node, context) {
  const { helpers } = context
  if (node.type === NodeTypes.ELEMENT) {
    helpers.add(CREATE_ELEMENT_VNODE)
  }
}
