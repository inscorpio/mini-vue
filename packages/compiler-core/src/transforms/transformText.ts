import { NodeTypes } from '../ast'
import { isText } from '../utils'

export function transformText(node) {
  if (node.type === NodeTypes.ELEMENT) {
    const { children } = node
    for (let i = 0; i < children.length; i++) {
      const child = children[i]
      let compoundNode
      if (isText(child)) {
        for (let j = i + 1; j < children.length; j++) {
          const next = children[j]
          if (isText(next)) {
            compoundNode = {
              type: NodeTypes.COMPOUND_EXPRESSION,
              children: [
                child,
                ' + ',
                next,
              ],
            }
            children.splice(j, 1)
            j--
          }
          else {
            break
          }
        }
        if (compoundNode) {
          children[i] = compoundNode
        }
      }
    }
  }
}
