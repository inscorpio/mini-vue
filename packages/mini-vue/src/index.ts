export * from '@mini-vue/runtime-core'
