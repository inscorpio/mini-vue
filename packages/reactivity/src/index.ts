export {
  reactive,
  readonly,
  shallowReadonly,
  isReactive,
  isReadonly,
  isProxy,
} from './reactive'

export {
  ref,
  isRef,
  unref,
  proxyRefs,
} from './ref'
