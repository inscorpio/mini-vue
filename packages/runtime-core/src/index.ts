export * from './createApp'
export * from './h'
export { getCurrentInstance } from './component'
export {
  provide,
  inject,
} from './apiInject'
