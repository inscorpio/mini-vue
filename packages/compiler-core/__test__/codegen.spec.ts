import { describe, expect, it } from 'vitest'
import { baseParse } from '../src/parse'
import { transform } from '../src/transform'
import { generate } from '../src/codegen'
import { transformExpression } from '../src/transforms/transformExpression'
import { transformElement } from '../src/transforms/transformElement'
import { transformText } from '../src/transforms/transformText'

describe('codegen', () => {
  it('should return a render function with text', () => {
    const ast = baseParse('hello world')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('should return a render function with interpolation', () => {
    const ast = baseParse('{{message}}')
    transform(ast, { nodeTransforms: [transformExpression] })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('should return a render function only with element', () => {
    const ast = baseParse('<div></div>')
    transform(ast, { nodeTransforms: [transformElement] })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('should return a render function with element nested with text', () => {
    const ast = baseParse('<div>hello world</div>')
    transform(ast, { nodeTransforms: [transformElement] })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('should return a render function with compound node', () => {
    const ast = baseParse('<div>hello {{message}}</div>')
    transform(ast, { nodeTransforms: [transformElement, transformText, transformExpression] })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('should return a render function with compound node 2', () => {
    const ast = baseParse('<div><span>hello </span>{{message}}</div>')
    transform(ast, { nodeTransforms: [transformElement, transformText, transformExpression] })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})
