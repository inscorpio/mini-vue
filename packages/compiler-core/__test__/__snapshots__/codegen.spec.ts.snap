// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > should return a render function only with element 1`] = `
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx) {
  return _createElementVNode(\\"div\\")
}"
`;

exports[`codegen > should return a render function with compound node 1`] = `
"const { createElementVNode: _createElementVNode, toDisplayString: _toDisplayString } = Vue
return function render(_ctx) {
  return _createElementVNode(\\"div\\", null, \\"hello \\" + _toDisplayString(_ctx.message))
}"
`;

exports[`codegen > should return a render function with compound node 2 1`] = `
"const { createElementVNode: _createElementVNode, toDisplayString: _toDisplayString } = Vue
return function render(_ctx) {
  return _createElementVNode(\\"div\\", null, [_createElementVNode(\\"span\\", null, \\"hello \\"), _toDisplayString(_ctx.message)])
}"
`;

exports[`codegen > should return a render function with element nested with text 1`] = `
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx) {
  return _createElementVNode(\\"div\\", null, \\"hello world\\")
}"
`;

exports[`codegen > should return a render function with interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx) {
  return _toDisplayString(_ctx.message)
}"
`;

exports[`codegen > should return a render function with text 1`] = `
"return function render(_ctx) {
  return \\"hello world\\"
}"
`;
